mycore-base/src/main/java/org/mycore/common/log4j2/filter/MCRUniqueFilter.java

```diff
@@ -0,0 +1,85 @@
+package org.mycore.common.log4j2.filter;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+import org.apache.logging.log4j.Marker;
+import org.apache.logging.log4j.core.Filter;
+import org.apache.logging.log4j.core.LogEvent;
+import org.apache.logging.log4j.core.config.Node;
+import org.apache.logging.log4j.core.config.plugins.Plugin;
+import org.apache.logging.log4j.core.config.plugins.PluginAttribute;
+import org.apache.logging.log4j.core.config.plugins.PluginFactory;
+import org.apache.logging.log4j.core.filter.AbstractFilter;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+
+/**
+ * Filters duplicate instances of logging events that have the same parameters for a given format.
+ * Only filters logging events that have a given marker, configures by the <code>targetMarker</code> configuration
+ * value. Uses an LRU cache to only filters up to a maximum amount of parameter sets per format. The cache size per
+ * format con be configured with the <code>cacheSizePerFormat</code> configuration value; defaults to
+ * <code>1000</code>.
+ */
+@Plugin(name = "MCRUniqueFilter", category = Node.CATEGORY, elementType = Filter.ELEMENT_TYPE)
+public class MCRUniqueFilter extends AbstractFilter {
+
+    private final String targetMarker;
+
+    private final int cacheSizePerFormat;
+
+    private final Map<String, Set<List<Object>>> cacheByFormat = new HashMap<>();
+
+    private MCRUniqueFilter(String targetMarker, int cacheSizePerFormat, Result onMatch, Result onMismatch) {
+        super(onMatch, onMismatch);
+        this.targetMarker = targetMarker;
+        this.cacheSizePerFormat = cacheSizePerFormat;
+    }
+
+    @Override
+    public Result filter(LogEvent event) {
+        if (event.getMessage() instanceof ParameterizedMessage message) {
+            Marker marker = event.getMarker();
```

Comment:
It would look like, if i understand the log4j api correct:
```
   <logger name="org.mycore.common.xml.MCRURIResolver" >
      <Filters>
        <MarkerFilter marker="tryResolveXML" onMatch="NEUTRAL" onMismatch="ACCEPT"/>
         <MCRUniqueFilter />
      </Filters>
    </logger>
```
I think to match the log4j logic the MCRUniqueFilter should return the onMatch and the onMismatch fields.

Reply:
Background: I've added the marker to reduce the resource load. Each "message format string" will create a cache for the sets of parameter values used in combination with that "message format string", so I'd like to have a mechanism to restrict the amount of caches to the base minimum.

Log4js [documentation](https://logging.apache.org/log4j/2.x/manual/filters.html#CompositeFilter) on the composite filter is quite sparse.

> The CompositeFilter provides a way to specify more than one filter. It is added to the configuration as a filters element and contains other filters to be evaluated. The filters element accepts no parameters. 

The class comment on the [implementation](https://logging.apache.org/log4j/log4j-2.17.2/log4j-core/apidocs/src-html/org/apache/logging/log4j/core/filter/CompositeFilter.html) is not helpful either.

> Composes and invokes one or more filters.

Looking at the implementation shows that the first filter result that yields a non-neutral result is used.

```
@Override
public Result filter(final LogEvent event) {
    Result result = Result.NEUTRAL;
    for (int i = 0; i < filters.length; i++) {
        result = filters[i].filter(event);
        if (result == Result.ACCEPT || result == Result.DENY) {
            return result;
        }
    }
    return result;
}
```

So, I think your suggestion should work.